### PR TITLE
Improved m-separation property

### PR DESCRIPTION
**Pull request by jaron-lee**
This PR improves the m-separation property by adapting the existing networkx.algorithms.d_separation code to work for mixed-edge graphs.

### CODE REVIEW

File: networkx/algorithms/causal/m_separation.py [hunk 1]
```diff
@@ -1,21 +1,35 @@
+from copy import deepcopy
+from collections import deque
+
 import networkx as nx
+from networkx.utils import UnionFind
+
+import graphs
```
Comment (adam2392) [inline networkx/algorithms/causal/m_separation.py]: Don't need this anymore

File: networkx/algorithms/causal/m_separation.py [hunk 2]
```diff
@@ -1,21 +1,35 @@
...
 
 __all__ = ["m_separated"]
 
 
+def markov_blanket(
+    G, x, bidirected_edge_name="bidirected", directed_edge_name="directed"
+):
+    pass
+
+
```
Comment (adam2392) [inline networkx/algorithms/causal/m_separation.py]: Is this needed for anything?

File: networkx/algorithms/causal/m_separation.py [hunk 3]
```diff
@@ -1,21 +1,35 @@
...
+    undirected graph composed from the directed and bidirected edges amongst the remaining nodes is
+    created. Then, x is independent of y given z if x is disconnected from y in this new graph.
 
-    It first converts all bidirected edges into explicit unobserved
-    confounding nodes in an explicit ``networkx.DiGraph``, which then
-    calls ``networkx.algorithms.d_separated`` to determine d-separation.
-    This inherently increases the runtime cost if there are many
-    bidirected edges, because many nodes must be added.
+    [1] Darwiche, A.  (2009).  Modeling and reasoning with Bayesian networks.
+       Cambridge: Cambridge University Press.
```
Comment (adam2392) [inline networkx/algorithms/causal/m_separation.py]: In networkx, they cite things with the patterns seen in this function: https://github.com/networkx/networkx/blob/d2ceeb69ab0580a84a3b13266e6bd32e99da0232/networkx/algorithms/d_separation.py#L331-L333

Comment (adam2392) [reply networkx/algorithms/causal/m_separation.py]: So a separate `References` section

File: networkx/algorithms/causal/m_separation.py [hunk 4]
```diff
@@ -35,14 +49,14 @@ def m_separated(
 
     See Also
     --------
-    d_separated
+    networkx.algorithms.d_separation.d_separated
```
Comment (adam2392) [inline networkx/algorithms/causal/m_separation.py]:
```suggestion
    d_separated
```

File: networkx/algorithms/causal/m_separation.py [hunk 5]
```diff
@@ -35,14 +49,14 @@ def m_separated(
...
 
     Notes
     -----
     This wraps the networkx implementation, which only allows DAGs. Since
     ``ADMG`` is not represented.
     """
-    if not isinstance(G, nx.MixedEdgeGraph):
+    if not isinstance(G, graphs.MixedEdgeGraph):
```
Comment (adam2392) [inline networkx/algorithms/causal/m_separation.py]:
```suggestion
    if not isinstance(G, nx.MixedEdgeGraph):
```

File: networkx/algorithms/causal/tests/test_m_separation.py [hunk 6]
```diff
@@ -1,25 +1,29 @@
...
 
     # error should be raised if someone does not use a MixedEdgeGraph
-    with pytest.raises(
-        RuntimeError, match="m-separation should only be run on a MixedEdgeGraph"
-    ):
-        nx.m_separated(digraph, {0}, {1}, set())
+    with pytest.raises(RuntimeError, match="m-separation should only be run on a MixedEdgeGraph"):
+        m_separated(digraph, {0}, {1}, set())
 
     print(G.edges())
```
Comment (adam2392) [inline networkx/algorithms/causal/tests/test_m_separation.py]:
```suggestion
```

File: networkx/algorithms/causal/tests/test_m_separation.py [hunk 7]
```diff
@@ -1,25 +1,29 @@
...
     # basic d-separation statements based on blocking paths should work
-    assert not nx.m_separated(G, {0}, {3}, set())
-    assert nx.m_separated(G, {0}, {3}, {1})
+    assert not m_separated(G, {0}, {3}, set())
```
Comment (adam2392) [inline networkx/algorithms/causal/tests/test_m_separation.py]:
```suggestion
    assert not nx.m_separated(G, {0}, {3}, set())
```

File: networkx/algorithms/causal/tests/test_m_separation.py [hunk 8]
```diff
@@ -1,25 +1,29 @@
...
+    assert m_separated(G, {0}, {3}, {1})
```
Comment (adam2392) [inline networkx/algorithms/causal/tests/test_m_separation.py]:
```suggestion
    assert nx.m_separated(G, {0}, {3}, {1})
```

File: networkx/algorithms/causal/tests/test_m_separation.py [hunk 9]
```diff
@@ -1,25 +1,29 @@
...
 
     # conditioning on a collider via bidirected edge opens the collider
-    assert not nx.m_separated(G, {0}, {3}, {2})
-    assert nx.m_separated(G, {0}, {3}, {1, 2})
+    assert not m_separated(G, {0}, {3}, {2})
+    assert m_separated(G, {0}, {3}, {1, 2})
```
Comment (adam2392) [inline networkx/algorithms/causal/tests/test_m_separation.py]:
```suggestion
    assert not nx.m_separated(G, {0}, {3}, {2})
    assert nx.m_separated(G, {0}, {3}, {1, 2})
```

File: networkx/algorithms/causal/tests/test_m_separation.py [hunk 10]
```diff
@@ -1,25 +1,29 @@
...
+
+    # conditioning on a descendant of a collider via bidirected edge opens the collider
+    assert not m_separated(G, {0}, {3}, {4})
```
Comment (adam2392) [inline networkx/algorithms/causal/tests/test_m_separation.py]:
```suggestion
    assert not nx.m_separated(G, {0}, {3}, {4})
```

File: networkx/algorithms/causal/m_separation.py [hunk 11]
```diff
@@ -1,21 +1,35 @@
...
     G, x, y, z, bidirected_edge_name="bidirected", directed_edge_name="directed"
 ):
     """Check m-separation among 'x' and 'y' given 'z' in mixed-edge causal graph G.
 
-    This algorithm wraps ``networkx.algorithms.d_separated``, but
-    allows one to pass in a ``ADMG`` instance instead.
+    This algorithm adapts the linear time algorithm presented in [1] currently implemeted in
+    networkx.algorithms.d_separation to work for mixed-edge causal graphs.
+
+    This algorithm first obtains the ancestral subgraph of x | y | z which only requires knowledge of
```
Comment (adam2392) [inline networkx/algorithms/causal/m_separation.py:18]: What does `x | y | z` mean here? Is that the union of only those nodes and gets the subgraph consisting only of the ancestors of that union of nodes?

**Review by adam2392 — COMMENTED** [review]
Pretty slick!

I didn't realize the way I implemented it was so inefficient. Some like coding style stuff that networkx will want to see, but otw LGTM!

File: networkx/algorithms/causal/m_separation.py (hunk 11 above)
Comment (jaron-lee) [reply networkx/algorithms/causal/m_separation.py:18]: Yep that's right. I would write X \cup Y \cup Z in latex, this particular chunk is actually copied verbatim from the current d-sep implementation

File: networkx/algorithms/causal/m_separation.py (hunk 2 above)
Comment (jaron-lee) [reply networkx/algorithms/causal/m_separation.py]: nope oops, i'll get rid of it

File: networkx/algorithms/causal/m_separation.py (hunk 3 above)
Comment (jaron-lee) [reply networkx/algorithms/causal/m_separation.py]: Yep copied the pattern

File: networkx/algorithms/causal/m_separation.py (hunk 4 above)
Comment (jaron-lee) [reply networkx/algorithms/causal/m_separation.py]: Noted

File: networkx/algorithms/causal/m_separation.py (hunk 5 above)
Comment (jaron-lee) [reply networkx/algorithms/causal/m_separation.py]: Noted

File: networkx/algorithms/causal/tests/test_m_separation.py (hunk 6 above)
Comment (jaron-lee) [reply networkx/algorithms/causal/tests/test_m_separation.py]: Agh I always leave these print statements, bad habit

**Comment by jaron-lee** [comment]
> 

Yeah so I looked more closely at it and the most expensive part of the algorithm is finding the weakly connected components in a DAG (or alternatively the connected components of a undirected graph) and this is O(|V| + |E|). In this implementation |V| is the number of nodes in original graph and |E| will be the number of directed and bidirected edges in the original graph. The advantage of this implementation is that we keep the smaller since in the previous proposal for each bidirected edge we will add one more node and one more edge.

File: networkx/algorithms/causal/m_separation.py [hunk 12]
```diff
@@ -52,27 +64,58 @@ def m_separated(
         for edge_type in [bidirected_edge_name, directed_edge_name]
     ):
         raise RuntimeError(
-            f"m-separation only works on nx with directed and bidirected edges. "
+            f"m-separation only works on graphs with directed and bidirected edges. "
             f"Your graph passed in has the following edge types: {G.edge_types}, whereas "
             f"the function is expecting directed edges named {directed_edge_name} and "
             f"bidirected edges named {bidirected_edge_name}."
         )
+    union_xyz = x.union(y).union(z)
```
Comment (adam2392) [inline networkx/algorithms/causal/m_separation.py:75]: Similar to `d_separated`, would it make sense to check the acyclic condition of the directed edges?

E.g.
```
    if not nx.is_directed_acyclic_graph(G.get_graphs(directed_edge_name)):
        raise nx.NetworkXError("directed edge graph should be directed acyclic")
```

**Review by adam2392 — APPROVED** [review]
One last comment, and then I think there's a style issue, which can be fixed and this is good to merge. If you run `black .` and `isort .`, or install the pre-commit hooks for networkx, it should fix the errors. 

https://github.com/networkx/networkx/blob/main/CONTRIBUTING.rst

```
black....................................................................Failed
[29](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:30)
- hook id: black
[30](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:31)
- files were modified by this hook
[31](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:32)

[32](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:33)
reformatted networkx/algorithms/causal/m_separation.py
[33](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:34)
reformatted networkx/algorithms/causal/tests/test_m_separation.py
[34](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:35)

[35](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:36)
All done! ✨ 🍰 ✨
[36](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:37)
2 files reformatted, 620 files left unchanged.
[37](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:38)

[38](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:39)
pyupgrade................................................................Passed
[39](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:40)
blacken-docs.............................................................Passed
[40](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:41)
isort (python)...........................................................Failed
[41](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:42)
- hook id: isort
[42](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:43)
- files were modified by this hook
[43](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:44)

[44](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:45)
Fixing /home/runner/work/networkx/networkx/networkx/algorithms/causal/m_separation.py
[45](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:46)
Fixing /home/runner/work/networkx/networkx/networkx/algorithms/causal/tests/test_m_separation.py
[46](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:47)
Skipped 50 files
[47](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:48)

[48](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:49)
prettier.................................................................Passed
[49](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:50)
pre-commit hook(s) made changes.
[50](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:51)
If you are seeing this message in CI, reproduce locally with: `pre-commit run --all-files`.
[51](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:52)
To run `pre-commit` as part of git workflow, use `pre-commit install`.
[52](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:53)
All changes made by hooks:
[53](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:54)
diff --git a/networkx/algorithms/causal/m_separation.py b/networkx/algorithms/causal/m_separation.py
[54](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:55)
index 2dee20b..cdca12f 100644
[55](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:56)
--- a/networkx/algorithms/causal/m_separation.py
[56](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:57)
+++ b/networkx/algorithms/causal/m_separation.py
[57](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:58)
@@ -1,10 +1,9 @@
[58](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:59)
-from copy import deepcopy
[59](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:60)
 from collections import deque
[60](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:61)
+from copy import deepcopy
[61](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:62)
 
[62](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:63)
 import networkx as nx
[63](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:64)
 from networkx.utils import UnionFind
[64](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:65)
 
[65](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:66)
-
[66](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:67)
 __all__ = ["m_separated"]
[67](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:68)
 
[68](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:69)
 
[69](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:70)
@@ -80,16 +79,12 @@ def m_separated(
[70](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:71)
     # get bidirected edges subgraph
[71](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:72)
     G_bidirected = nx.Graph()
[72](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:73)
     G_bidirected.add_nodes_from((n, deepcopy(d)) for n, d in G.nodes.items())
[73](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:74)
-    G_bidirected.add_edges_from(
[74](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:75)
-        G.get_graphs(edge_type=bidirected_edge_name).edges
[75](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:76)
-    )
[76](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:77)
+    G_bidirected.add_edges_from(G.get_graphs(edge_type=bidirected_edge_name).edges)
[77](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:78)
 
[78](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:79)
     # get ancestral subgraph of x | y | z by removing leaves in directed graph that are not in x | y | z
[79](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:80)
 
[80](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:81)
     # until no more leaves can be removed.
[81](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:82)
-    leaves = deque(
[82](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:83)
-        [n for n in G_directed.nodes if G_directed.out_degree[n] == 0]
[83](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:84)
-    )
[84](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:85)
+    leaves = deque([n for n in G_directed.nodes if G_directed.out_degree[n] == 0])
[85](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:86)
     while len(leaves) > 0:
[86](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:87)
         leaf = leaves.popleft()
[87](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:88)
         if leaf not in union_xyz:
[88](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:89)
diff --git a/networkx/algorithms/causal/tests/test_m_separation.py b/networkx/algorithms/causal/tests/test_m_separation.py
[89](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:90)
index 0bda958..58faafc 100644
[90](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:91)
--- a/networkx/algorithms/causal/tests/test_m_separation.py
[91](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:92)
+++ b/networkx/algorithms/causal/tests/test_m_separation.py
[92](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:93)
@@ -1,6 +1,7 @@
[93](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:94)
-import networkx as nx
[94](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:95)
 import pytest
[95](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:96)
 
[96](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:97)
+import networkx as nx
[97](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:98)
+
[98](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:99)
 
[99](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:100)
 def test_m_separation():
[100](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:101)
     digraph = nx.path_graph(4)
[101](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:102)
@@ -10,7 +11,9 @@ def test_m_separation():
[102](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:103)
     G = nx.MixedEdgeGraph([digraph, bigraph], ["directed", "bidirected"])
[103](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:104)
 
[104](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:105)
     # error should be raised if someone does not use a MixedEdgeGraph
[105](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:106)
-    with pytest.raises(RuntimeError, match="m-separation should only be run on a MixedEdgeGraph"):
[106](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:107)
+    with pytest.raises(
[107](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:108)
+        RuntimeError, match="m-separation should only be run on a MixedEdgeGraph"
[108](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:109)
+    ):
[109](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:110)
         nx.m_separated(digraph, {0}, {1}, set())
[110](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:111)
 
[111](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:112)
     # basic d-separation statements based on blocking paths should work
[112](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:113)
Error: Process completed with exit code 1.
```

File: networkx/algorithms/causal/m_separation.py (hunk 12 above)
Comment (adam2392) [reply networkx/algorithms/causal/m_separation.py:75]: Above this line that is. if the DAG check makes sense to run.

Comment (jaron-lee) [reply networkx/algorithms/causal/m_separation.py:75]: Yeah that makes sense, I can add that

**Comment by jaron-lee** [comment]
> One last comment, and then I think there's a style issue, which can be fixed and this is good to merge. If you run `black .` and `isort .`, or install the pre-commit hooks for networkx, it should fix the errors.
> 
> https://github.com/networkx/networkx/blob/main/CONTRIBUTING.rst
> 
> ```
> black....................................................................Failed
> [29](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:30)
> - hook id: black
> [30](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:31)
> - files were modified by this hook
> [31](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:32)
> 
> [32](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:33)
> reformatted networkx/algorithms/causal/m_separation.py
> [33](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:34)
> reformatted networkx/algorithms/causal/tests/test_m_separation.py
> [34](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:35)
> 
> [35](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:36)
> All done! ✨ 🍰 ✨
> [36](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:37)
> 2 files reformatted, 620 files left unchanged.
> [37](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:38)
> 
> [38](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:39)
> pyupgrade................................................................Passed
> [39](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:40)
> blacken-docs.............................................................Passed
> [40](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:41)
> isort (python)...........................................................Failed
> [41](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:42)
> - hook id: isort
> [42](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:43)
> - files were modified by this hook
> [43](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:44)
> 
> [44](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:45)
> Fixing /home/runner/work/networkx/networkx/networkx/algorithms/causal/m_separation.py
> [45](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:46)
> Fixing /home/runner/work/networkx/networkx/networkx/algorithms/causal/tests/test_m_separation.py
> [46](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:47)
> Skipped 50 files
> [47](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:48)
> 
> [48](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:49)
> prettier.................................................................Passed
> [49](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:50)
> pre-commit hook(s) made changes.
> [50](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:51)
> If you are seeing this message in CI, reproduce locally with: `pre-commit run --all-files`.
> [51](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:52)
> To run `pre-commit` as part of git workflow, use `pre-commit install`.
> [52](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:53)
> All changes made by hooks:
> [53](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:54)
> diff --git a/networkx/algorithms/causal/m_separation.py b/networkx/algorithms/causal/m_separation.py
> [54](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:55)
> index 2dee20b..cdca12f 100644
> [55](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:56)
> --- a/networkx/algorithms/causal/m_separation.py
> [56](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:57)
> +++ b/networkx/algorithms/causal/m_separation.py
> [57](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:58)
> @@ -1,10 +1,9 @@
> [58](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:59)
> -from copy import deepcopy
> [59](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:60)
>  from collections import deque
> [60](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:61)
> +from copy import deepcopy
> [61](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:62)
>  
> [62](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:63)
>  import networkx as nx
> [63](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:64)
>  from networkx.utils import UnionFind
> [64](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:65)
>  
> [65](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:66)
> -
> [66](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:67)
>  __all__ = ["m_separated"]
> [67](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:68)
>  
> [68](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:69)
>  
> [69](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:70)
> @@ -80,16 +79,12 @@ def m_separated(
> [70](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:71)
>      # get bidirected edges subgraph
> [71](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:72)
>      G_bidirected = nx.Graph()
> [72](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:73)
>      G_bidirected.add_nodes_from((n, deepcopy(d)) for n, d in G.nodes.items())
> [73](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:74)
> -    G_bidirected.add_edges_from(
> [74](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:75)
> -        G.get_graphs(edge_type=bidirected_edge_name).edges
> [75](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:76)
> -    )
> [76](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:77)
> +    G_bidirected.add_edges_from(G.get_graphs(edge_type=bidirected_edge_name).edges)
> [77](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:78)
>  
> [78](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:79)
>      # get ancestral subgraph of x | y | z by removing leaves in directed graph that are not in x | y | z
> [79](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:80)
>  
> [80](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:81)
>      # until no more leaves can be removed.
> [81](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:82)
> -    leaves = deque(
> [82](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:83)
> -        [n for n in G_directed.nodes if G_directed.out_degree[n] == 0]
> [83](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:84)
> -    )
> [84](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:85)
> +    leaves = deque([n for n in G_directed.nodes if G_directed.out_degree[n] == 0])
> [85](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:86)
>      while len(leaves) > 0:
> [86](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:87)
>          leaf = leaves.popleft()
> [87](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:88)
>          if leaf not in union_xyz:
> [88](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:89)
> diff --git a/networkx/algorithms/causal/tests/test_m_separation.py b/networkx/algorithms/causal/tests/test_m_separation.py
> [89](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:90)
> index 0bda958..58faafc 100644
> [90](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:91)
> --- a/networkx/algorithms/causal/tests/test_m_separation.py
> [91](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:92)
> +++ b/networkx/algorithms/causal/tests/test_m_separation.py
> [92](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:93)
> @@ -1,6 +1,7 @@
> [93](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:94)
> -import networkx as nx
> [94](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:95)
>  import pytest
> [95](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:96)
>  
> [96](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:97)
> +import networkx as nx
> [97](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:98)
> +
> [98](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:99)
>  
> [99](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:100)
>  def test_m_separation():
> [100](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:101)
>      digraph = nx.path_graph(4)
> [101](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:102)
> @@ -10,7 +11,9 @@ def test_m_separation():
> [102](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:103)
>      G = nx.MixedEdgeGraph([digraph, bigraph], ["directed", "bidirected"])
> [103](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:104)
>  
> [104](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:105)
>      # error should be raised if someone does not use a MixedEdgeGraph
> [105](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:106)
> -    with pytest.raises(RuntimeError, match="m-separation should only be run on a MixedEdgeGraph"):
> [106](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:107)
> +    with pytest.raises(
> [107](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:108)
> +        RuntimeError, match="m-separation should only be run on a MixedEdgeGraph"
> [108](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:109)
> +    ):
> [109](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:110)
>          nx.m_separated(digraph, {0}, {1}, set())
> [110](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:111)
>  
> [111](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:112)
>      # basic d-separation statements based on blocking paths should work
> [112](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:113)
> Error: Process completed with exit code 1.
> ```

Yep I will get onto those pre-commit hooks

**Comment by jaron-lee** [comment]
I fixed up all the changes and tidied up the commit history, should be good to go.

**Comment by adam2392** [comment]
> I fixed up all the changes and tidied up the commit history, should be good to go.

Awesome! Thanks a lot Jaron. I'll merge as soon as the CI turns green.

**Comment by adam2392** [comment]
Pushed a few fixes on the unit test, mypy and style.

**Pull request merged**